### ISSUE TITLE
')spool out.txt' does not write the output of a FriCAS command in the interpreter

### ISSUE THREAD

**Issue by gvanuxem** [fricas/fricas#167]
From the )help command:

```
(1) -> )help spool
====================================================================
A.23.  )spool
====================================================================

User Level Required:  interpreter

Command Syntax:

  - )spool [fileName]
  - )spool

Command Description:

This command is used to save (spool) all FriCAS input and output into a file,
called a spool file. You can only have one spool file active at a time. To
start spool, issue this command with a filename.
[snip]
```

But:
```
Checking for foreign routines
FRICAS="/home/greg/.local/lib/fricas/target/x86_64-linux-gnu"
spad-lib="/home/greg/.local/lib/fricas/target/x86_64-linux-gnu/lib/libspad.so"
foreign routines found
openServer result 0
                       FriCAS Computer Algebra System
              Version: FriCAS 2023-06-17 built with sbcl 2.4.3
                 Timestamp: jeu. 11 avril 2024 03:22:23 CEST
-----------------------------------------------------------------------------
   Issue )copyright to view copyright notices.
   Issue )summary for a summary of useful system commands.
   Issue )quit to leave FriCAS and return to shell.
-----------------------------------------------------------------------------


(1) -> )spool out.txt

(1) -> erfi(3)

   (1)  erfi(3)
                                                    Type: Expression(Integer)
(2) -> )spool

(2) -> )sys cat out.txt

(1) -> erfi(3)
(2) -> )spool
```

**Comment by oldk1331** [comment]
With the help of fricas0-repo, I bisect to commit eb8d9a0e08f9cc56f8866f46d4c82d8c33e809eb.

**Comment by gvanuxem** [comment]
> With the help of fricas0-repo, I bisect to commit [eb8d9a0](https://github.com/fricas/fricas/commit/eb8d9a0e08f9cc56f8866f46d4c82d8c33e809eb).

I wonder what is `fricas0-repo`, running 'find -name \*fricas0-repo\*' in the FriCAS source tree returns nothing.

**Comment by hemmecke** [comment]
Maybe you know this guy... ;-)
https://github.com/oldk1331
I think when you study his fricas0 repo, you will find out what it is about.

**Comment by oldk1331** [comment]
It's https://github.com/oldk1331/fricas0-repo/, which contains pre-compiled Lisp files of 1800 commits of FriCAS.

So you can easily run any past version in an instant, together with https://github.com/oldk1331/fricas0/blob/master/fricas.lisp (actually a variant of it to be compatible with older versions.)

**Comment by oldk1331** [comment]
OK, I have found the problematic part is:

````
(defun |mkOutputConsoleStream| () (CONS NIL *standard-output*))
````

Changing back to the old way works:

````
(defun |mkOutputConsoleStream| () (CONS NIL (make-synonym-stream '*standard-output*)))
````

**Comment by hebisch** [comment]
On Fri, Apr 12, 2024 at 07:12:06PM -0700, Qian Yun wrote:
> OK, I have found the problematic part is:
> 
> ````
> (defun |mkOutputConsoleStream| () (CONS NIL *standard-output*))
> ````
> 
> Changing back to the old way works:
> 
> ````
> (defun |mkOutputConsoleStream| () (CONS NIL (make-synonym-stream '*standard-output*)))
> ````

I would very much prefer different sulution.  The whole point of the
change was to decouple FriCAS streams from Lisp streams.

')spool' should be reimplements, using Lisp 'DRIBBLE' simply is
wrong.

-- 
                              Waldek Hebisch

**Comment by hebisch** [comment]
On Fri, Apr 12, 2024 at 07:12:06PM -0700, Qian Yun wrote:
> OK, I have found the problematic part is:
> 
> ````
> (defun |mkOutputConsoleStream| () (CONS NIL *standard-output*))
> ````
> 
> Changing back to the old way works:
> 
> ````
> (defun |mkOutputConsoleStream| () (CONS NIL (make-synonym-stream '*standard-output*)))
> ````

I would very much prefer different sulution.  The whole point of the
change was to decouple FriCAS streams from Lisp streams.

')spool' should be reimplemented, using Lisp 'DRIBBLE' simply is
wrong.

-- 
                              Waldek Hebisch

**Comment by hebisch** [comment]
On Fri, Apr 12, 2024 at 04:12:38AM -0700, Grégory Vanuxem wrote:
> >From the )help command:
> 
> ```
> (1) -> )help spool
> ====================================================================
> A.23.  )spool
> ====================================================================
> 
> User Level Required:  interpreter
> 
> Command Syntax:
> 
>   - )spool [fileName]
>   - )spool
> 
> Command Description:
> 
> This command is used to save (spool) all FriCAS input and output into a file,
> called a spool file. You can only have one spool file active at a time. To
> start spool, issue this command with a filename.
> [snip]
> ```
> 
> But:
> ```
> Checking for foreign routines
> FRICAS="/home/greg/.local/lib/fricas/target/x86_64-linux-gnu"
> spad-lib="/home/greg/.local/lib/fricas/target/x86_64-linux-gnu/lib/libspad.so"
> foreign routines found
> openServer result 0
>                        FriCAS Computer Algebra System
>               Version: FriCAS 2023-06-17 built with sbcl 2.4.3
>                  Timestamp: jeu. 11 avril 2024 03:22:23 CEST
> -----------------------------------------------------------------------------
>    Issue )copyright to view copyright notices.
>    Issue )summary for a summary of useful system commands.
>    Issue )quit to leave FriCAS and return to shell.
> -----------------------------------------------------------------------------
> 
> 
> (1) -> )spool out.txt
> 
> (1) -> erfi(3)
> 
>    (1)  erfi(3)
>                                                     Type: Expression(Integer)
> (2) -> )spool
> 
> (2) -> )sys cat out.txt
> 
> (1) -> erfi(3)
> (2) -> )spool
> ```

This is known problem.  It was discussed in January (and possibly earlier)
in FriCAS mailing list.  If your issue is that ')spool' is documented,
than this is easy to fix: we can easily remove documentation.

-- 
                              Waldek Hebisch

**Comment by oldk1331** [comment]
I think this functionality can be implemented by redefine a few functions: `serverReadLine`, `printexp`, `TERPRI`, etc.

When checking that "spool" is active, these functions will write to that spool stream as well.

I think this can capture most of the input/output.  This method can't capture when other Lisp functions are used for IO, for example "format".

**Comment by gvanuxem** [comment]
> This is known problem. It was discussed in January (and possibly earlier) in FriCAS mailing list. If your issue is that ')spool' is documented, than this is easy to fix: we can easily remove documentation.
> […](#)
> -- Waldek Hebisch

No, it's just to show the documentation for someone who does not it. And to show that the system command `)spool` does not comply with what is stated in the documentation. About the fact it was discussed, I did not see this discussion, sometimes I disconnect phone(s) and computer(s) for one or more weeks, and usually I am unable to read all what I have missed.

To add to this, removing (part of ? ) the documentation would be an error for me. My point is that it is not what can be expected of this command. `)history )save | )restore` etc. already do the job to store input commands, here this command is useless for output of called Spad operations. When I write an input script, I like sometimes to read the output it in an editor, I can search in it for errors for example (ex. I wrote an input file with 87 tests, this is handy). I also use sometimes this command to store the output of the `)show` command for knowing what a domain or a package exports, but here that's work as expected.

**Comment by hebisch** [comment]
On Fri, Apr 12, 2024 at 09:32:35PM -0700, Qian Yun wrote:
> I think this functionality can be implemented by redefine a few functions: `serverReadLine`, `printexp`, `TERPRI`, etc.

I principle yes.  Basically, the problem is that there are several
such functions and they are called in different contexts, so there
is some remaing work to sort this out.

> When checking that "spool" is active, these functions will write to that spool stream as well.
> 
> I think this can capture most of the input/output.  This method can't capture when other Lisp functions are used for IO, for example "format".

FORMAT for i/o should go away.  OK, there could be some messages during
low-level startup, but most calls to FORMAT should be removed.

-- 
                              Waldek Hebisch

**Comment by oldk1331** [comment]
It looks like a even better way to do "spool" is to do it at C level: we can extend the protocol a bit, so that "session" binary can write a copy of the IO to the spool file.

The only disadvantage is that we can't use this functionality with FRICASsys alone, (aka it has to run with sman.)

**Comment by gvanuxem** [comment]
See also this mail for differences between FriCAS, here, whether it is built with SBCL or Clozure CL:

https://groups.google.com/g/fricas-devel/c/aZJICdOO8QI
